### PR TITLE
Sidebar abstraction

### PR DESCRIPTION
**Pull request by readeral**
Example using hymn page

### CODE REVIEW

**Comment by readeral** [comment]
I have no idea how type checking works
so the type SidebarProps needs thoroughly looking through, as does propTypes and defaultProps - and also how I'm passing the props through on hymn/index in the first place

File: pages/hymn/[id]/index.tsx [hunk 1]
```diff
@@ -78,64 +36,7 @@ const Song: NextPage<SongProps> = ({id}) => {
...
-          )}
-
-          {tune?.musicCopyright && (
-            <>
-              <Styled.h3>Copyright (music)</Styled.h3>
-              <Styled.p>{tune.musicCopyright.name || '-'}</Styled.p>
-            </>
-          )}
-        </Box>
+        <Sidebar {...data?.hymnById} data={data} />
```
Comment (BarryThePenguin) [inline pages/hymn/[id]/index.tsx]: Maybe something more like

```
<Box>
  {data?.hymnById && <Sidebar {...data.hymnById} />
</Box>
```

That keeps the styles seperate from the data. Plus, the Sidebar will only render when we have the hymn data

The thing with apollo is it does two renders on the server. The first pass is while it's fetching data, and `data` will by `null`. The second pass is after the data has been received, where `data` will have the response. This is also why there's a `loading` and `error` returned from the apollo hooks. Think of them as being 3 possible states, either loading, error, or data

File: components/sidebar.tsx [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  data?: Object;
+};
+
+const Sidebar: FC<SidebarProps> = ({
+  author,
+  scripture,
+  tune,
+  copyright,
+  alternateTunes,
+  data
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```suggestion
 ...props
```

File: components/sidebar.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+}) => {
+  let files = data?.hymnById?.files || [];
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]: let files = props.files || []

File: pages/hymn/[id]/index.tsx (hunk 1 above)
Comment (readeral) [reply pages/hymn/[id]/index.tsx]: Cool. Made that change.

File: components/sidebar.tsx (hunk 2 above)
Comment (readeral) [reply components/sidebar.tsx]: or `files: fileList`
with below being `let files = fileList || []`

File: components/sidebar.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  files?: Array;
+  author?: any;
+  scripture?: String;
+  tune?: Object;
+  alternateTunes?: Array;
+  copyright?: Object;
+  data?: Object;
+};
+
+const Sidebar: FC<SidebarProps> = ({
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```ts
import {SearchResult} from './queries';

const Sidebar: FC<SearchResult> = ({
```

Should get you there 👍 then you can ditch `SidebarProps`

`SearchResult` contains all the possible type information for `Hymn | Prayer | Liturgy | Scripture`

File: components/sidebar.tsx (hunk 3 above)
Comment (readeral) [reply components/sidebar.tsx]: see above

File: components/sidebar.tsx (hunk 2 above)
Comment (BarryThePenguin) [reply components/sidebar.tsx]: much of a muchness

File: components/sidebar.tsx (hunk 4 above)
Comment (readeral) [reply components/sidebar.tsx]: Done

Comment (readeral) [reply components/sidebar.tsx]: "Property 'author' does not exist on type 'PropsWithChildren<SearchResult>'"

Comment (readeral) [reply components/sidebar.tsx]: (does this for all props)

File: components/sidebar.tsx [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+SidebarTune.propTypes = {
+  composer: PropTypes.object,
+  metre: PropTypes.shape({
+    metre: PropTypes.string
+  })
+};
+
+SidebarTune.defaultProps = {
+  composer: null,
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```suggestion
  composer: undefined,
```

File: components/sidebar.tsx [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    </>
+  );
+};
+
+SidebarCopyright.propTypes = {
+  name: PropTypes.string
+};
+
+SidebarCopyright.defaultProps = {
+  name: 'Public Domain'
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]: Is that right? I have no idea 😅 

Keep in mind there is also the `<Styled.p>{name || '-'}</Styled.p>` part..

File: components/sidebar.tsx [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+};
+
+SidebarMusicCopyright.propTypes = {
+  musicCopyright: PropTypes.shape({
+    name: PropTypes.string
+  })
+};
+
+SidebarMusicCopyright.defaultProps = {
+  musicCopyright: {}
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```suggestion
  musicCopyright: undefined
```

File: components/sidebar.tsx (hunk 5 above)
Comment (readeral) [reply components/sidebar.tsx]: assuming this goes for all

File: components/sidebar.tsx [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+SidebarCopyright.defaultProps = {
+  name: 'Public Domain'
+};
+
+const SidebarMusicCopyright: FC<Tune> = ({musicCopyright}) => {
+  return (
+    <>
+      <Styled.h3>Copyright (music)</Styled.h3>
+      <Styled.p>{musicCopyright.name || '-'}</Styled.p>
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```suggestion
      <Styled.p>{props?.musicCopyright?.name || '-'}</Styled.p>
```

File: components/sidebar.tsx [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+const Sidebar = ({children}) => {
+  return <Box sx={{marginRight: '40px'}}>{children}</Box>;
+};
+
+Sidebar.propTypes = {
+  children: PropTypes.any
+};
+
+Sidebar.defaultProps = {
+  children: null
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```suggestion
  children: undefined
```

File: components/sidebar.tsx [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+SidebarMusicCopyright.defaultProps = {
+  musicCopyright: {}
+};
+
+const Sidebar = ({children}) => {
+  return <Box sx={{marginRight: '40px'}}>{children}</Box>;
+};
+
+Sidebar.propTypes = {
+  children: PropTypes.any
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```suggestion
  children: PropTypes.node
```

File: components/sidebar.tsx [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+  musicCopyright: PropTypes.shape({
+    name: PropTypes.string
+  })
+};
+
+SidebarMusicCopyright.defaultProps = {
+  musicCopyright: {}
+};
+
+const Sidebar = ({children}) => {
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```suggestion
const Sidebar: FC = ({children}) => {
```

File: components/sidebar.tsx [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+export default Sidebar;
+export {
+  SidebarAuthor,
+  SidebarCopyright,
+  SidebarFiles,
+  SidebarMusicCopyright,
+  SidebarScripture,
+  SidebarTune
+};
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]: Can either do this, or.. `export const SidebarFoo = ...` all the things

File: components/sidebar.tsx (hunk 8 above)
Comment (readeral) [reply components/sidebar.tsx]: also `const SidebarMusicCopyright: FC<Tune> = props => {`?

File: pages/hymn/[id]/index.tsx [hunk 13]
```diff
@@ -75,68 +33,13 @@ const Song: NextPage<SongProps> = ({id}) => {
...
-          )}
-
-          {tune?.musicCopyright && (
-            <>
-              <Styled.h3>Copyright (music)</Styled.h3>
-              <Styled.p>{tune.musicCopyright.name || '-'}</Styled.p>
-            </>
-          )}
+        <Box sx={{marginRight: '40px'}}>
+          {data?.hymnById && <Sidebar {...data.hymnById} />}
```
Comment (BarryThePenguin) [inline pages/hymn/[id]/index.tsx]:
```suggestion
        <Sidebar>
          {data?.hymnById && (
            <>
              {files.length > 0 && <SidebarFiles files={files} />}
              {author && <SidebarAuthor {...author} />}
              {scripture && <SidebarScripture scripture={scripture} />}
              {tune && <SidebarTune {...tune} />}
              {copyright && <SidebarCopyright {...copyright} />}
              {tune?.musicCopyright && <SidebarMusicCopyright {...tune} />}
            </>
          )}
        </Sidebar>
```

😚👌

File: components/sidebar.tsx (hunk 8 above)
Comment (BarryThePenguin) [reply components/sidebar.tsx]: Ya 👍 there's lots of ways of doing it

File: components/sidebar.tsx (hunk 12 above)
Comment (readeral) [reply components/sidebar.tsx]: changed

File: components/sidebar.tsx (hunk 6 above)
Comment (readeral) [reply components/sidebar.tsx]: Probably not a safe assumption, no...

File: pages/hymn/[id]/index.tsx [hunk 14]
```diff
@@ -75,68 +33,24 @@ const Song: NextPage<SongProps> = ({id}) => {
         sx={{
           flexDirection: ['column-reverse', 'column-reverse', 'row'],
           // TODO: What should this value actually be?
-          gap: '2em'
+          justifyContent: 'flex-start'
         }}
       >
-        <Box>
-          {files.length > 0 && (
+        <Box sx={{marginRight: '40px'}}>
```
Comment (BarryThePenguin) [inline pages/hymn/[id]/index.tsx]: I think this <Box> is the same as the Sidebar?

File: pages/hymn/[id]/index.tsx [hunk 15]
```diff
@@ -13,58 +12,17 @@ import {defineAbilitiesFor} from '../../../lib/abilities';
...
-  if (tune?.file) {
-    files = files.concat(tune.file);
-  }
-
-  if (alternateTunes) {
-    files = files.concat(alternateTunes.map(x => x.file));
-  }
-
-  files = files.filter(Boolean);
+  const {hymnNumber, content, title} = data?.hymnById ?? {};
```
Comment (BarryThePenguin) [inline pages/hymn/[id]/index.tsx]: We should be able to revert these lines back now too 👍

File: components/sidebar.tsx [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+SidebarTune.propTypes = {
+  composer: PropTypes.object,
+  metre: PropTypes.shape({
+    metre: PropTypes.string
+  })
+};
+
+SidebarTune.defaultProps = {
+  composer: undefined,
+  metre: {}
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```suggestion
  metre: undefined
```

File: components/sidebar.tsx [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+
+Composer.defaultProps = {
+  name: undefined
+};
+
+export const SidebarFiles: FC<{
+  file: string;
+  files: Array;
+  alternateTunes: Array;
+}> = ({file, files: fileList, alternateTunes}) => {
```
Comment (BarryThePenguin) [inline components/sidebar.tsx]:
```
export const SidebarFiles: FC<{files: Asset[]}> = ({files}) => (
```

File: pages/hymn/[id]/index.tsx (hunk 14 above)
Comment (readeral) [reply pages/hymn/[id]/index.tsx]: fixed

**Pull request merged**